graphql_api/types/test_analytics/test_analytics.py

```diff
@@ -1,24 +1,318 @@
+    term: str | None = None,
+) -> TestResultConnection:
+    """
+    Function that retrieves aggregated information about all tests in a given repository, for a given time range, optionally filtered by branch name.
+    The fields it calculates are: the test failure rate, commits where this test failed, last duration and average duration of the test.
+
+    :param repoid: repoid of the repository we want to calculate aggregates for
+    :param branch: optional name of the branch we want to filter on, if this is provided the aggregates calculated will only take into account
+        test instances generated on that branch. By default branches will not be filtered and test instances on all branches wil be taken into
+        account.
+    :param interval: timedelta for filtering test instances used to calculate the aggregates by time, the test instances used will be
+        those with a created at larger than now - interval.
+    :param testsuites: optional list of testsuite names to filter by, this is done via a union
+    :param flags: optional list of flag names to filter by, this is done via a union so if a user specifies multiple flags, we get all tests with any
+        of the flags, not tests that have all of the flags
+    :returns: queryset object containing list of dictionaries of results
+
+    """
+    repo = Repository.objects.get(repoid=repoid)
+    if branch is None:
+        branch = repo.branch
+    interval = measurement_interval.value
+    validate(interval, ordering, ordering_direction, after, before, first, last)
+
+    table = get_results(repoid, branch, interval)
+
+    if table is None:
+        return TestResultConnection(
+            edges=[],
+            total_count=0,
+            page_info={},
+        )
+
+    if term:
+        table = table.filter(pl.col("name").str.starts_with(term))
+
+    if testsuites:
+        table = table.filter(
+            pl.col("testsuite").is_not_null() & pl.col("testsuite").is_in(testsuites)
+        )
+
+    if flags:
+        table = table.filter(
+            pl.col("flags").list.eval(pl.element().is_in(flags)).list.any()
+        )
+
+    match parameter:
+        case TestResultsFilterParameter.FAILED_TESTS:
+            table = table.filter(pl.col("total_fail_count") > 0)
+        case TestResultsFilterParameter.FLAKY_TESTS:
+            table = table.filter(pl.col("total_flaky_fail_count") > 0)
+        case TestResultsFilterParameter.SKIPPED_TESTS:
+            table = table.filter(
+                pl.col("total_skip_count") > 0 & pl.col("total_pass_count") == 0
+            )
+        case TestResultsFilterParameter.SLOWEST_TESTS:
+            table = table.filter(
+                pl.col("avg_duration") >= pl.col("avg_duration").quantile(0.95)
+            ).top_k(100, by=pl.col("avg_duration"))
+
+    total_count = table.height
+
+    def ordering_expression(cursor_value: CursorValue, is_forward: bool) -> pl.Expr:
+        if is_forward:
```

Comment:
what does forward imply here? It's also a little confusing that we use the value of is_forward and then change the value of is_forward in the same fn

Reply:
is_forward is just about determining the ordering of the table, but `ordering_expression` only reads is_forward, it's the outer function that sets the value and invokes `ordering_expression` so i ended up just moving it out